### ISSUE TITLE
Add effective constant propagation for Python

### ISSUE THREAD

**Issue by nbrahms** [returntocorp/semgrep#349]
**Is your feature request related to a problem? Please describe.**
We would like to treat the following programs as equivalent:

```python
x = 1
return x + 2
```

```python
return 1 + 2
```

Currently this works for string constants, but we'd like to extend to any literal variable that is not reassigned between assignment and use.

**Describe alternatives you've considered**
We can currently write a pattern like this:
```
$X = 1
...
return $X + 2
```

however, this does not guarantee that `$X` is not modified in between.

**Comment by aryx** [comment]
This already works for any literal constants, not just string constants, but the issue for Python is that there is no 'const' keyword so for Python we need a 'const' analysis, or in your case a dataflow-based constant propagation.

**Comment by aryx** [comment]
You want two different things here:
 - const "inference" for python, so a pattern like foo("bar") would match x = "bar"; foo(x)
   which is already available for Javascript and Go thanks to their const keyword
 - enforced that no modification happens to a variable in a '...'. In coccinelle we would write
   $X = 1
   ... when not $X = ...
   return $X + 2

But that's very different things.
Note that if we were associating to each variable a unique ID a la SSA, then your pattern $X = 1 ... return $X + 2 would automatically enforces that there is no assignment between, but I'm not sure we always want that behavior.

In any case those are two different things, so which one do you want? You simply want to write 
a pattern like return 1 + 2 and that it matches  x = 1; return x + 2? Do you want a real dataflow-based constant propagation so it handles correctly control flow? or just having a simple const analysis for python where we detect that a variable is assigned only once in the program is enough? (Python does not have 'const' keyword, but we can detect it if in practice it's assigned only once, and it's pretty easy to do).

**Comment by aryx** [comment]
the dataflow-based constant propagation is far more difficult to implement. There is already  some code for a generic control-flow and generic data-flow, but we need to plug them in sgrep and it's not battle-tested (it was done when the generic AST was supporting only Python and JS; now we have far more languages and so I'll need to update those generic controlf-flow and data-flow code.

**Comment by nbrahms** [comment]
@aryx In my opinion just detecting variable reassignment is likely sufficient.

For example, it's probably fine to treat `x` as "not constant" in this program:

```python
x = 1
if false:
    x = 2
return x + 2
```

**Comment by nbrahms** [comment]
Correct, I'd like to be able to just write:

```
return 1 + 2
```

and have it match the very first example described.

Is there a way to match any numeric literal, à la `"..."` for literal strings? Assuming for the sake of discussion we could do this with `...L`, I'd love to be able to match the example with

```
return ...L
```

**Comment by aryx** [comment]
In theory we could use typing, like 
return ($X: int_literal)
The problem is that sometimes there is a grammar ambiguity so we may have to invent another syntax.

Also where-python could be used in the yaml file.

**Comment by aryx** [comment]
This now works for Python globals.
We could activate this more generally to any variables, but this lead to some regressions right now.

**Comment by aryx** [comment]
Closing this as the situation has improved a bit (there was no const propagation at all before for Python).
Feel free to create another task if you want full const propagation for python, not just for the globals.
